internal/pkg/build/args/reader.go

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2019-2023, Sylabs Inc. All rights reserved.
+// Copyright (c) Contributors to the Apptainer project, established as
+//   Apptainer a Series of LF Projects LLC.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package args
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"regexp"
+
+	"github.com/samber/lo"
+)
+
+var buildArgsRegexp = regexp.MustCompile(`{{\s*(\w+)\s*}}`)
+
+func NewReader(src io.Reader, buildArgsMap map[string]string, defaultArgsMap map[string]string, consumedArgs *[]string) (io.Reader, error) {
```

Comment:
Would prefer it if it returned `consumedArgs` as a `[]string`, and the append was done by the caller.

If you don't change this, please make sure a function comment is added here the documents the side-effect on the passed in consumedArgs.

Reply:
My thinking here is that in the previous (now scrapped) implementation, the NewReader() constructor did not yet "know" which arguments were consumed; that was (and could only be) calculated in the course of Reader.Read(). The way this implementation works, NewReader() actually returns a plain-vanilla bytes.Reader, so the Read() method is "naive", and the knowledge of which arguments got consumed is already present in the constructor. This means that before, we got the consumed args info as part of the return value of Read(), and now, if we did it as a return value, we'd get it as part of the return value for the NewReader() constructor.

This implementation-dependence strikes me as a bad sign, from a software-design perspective. So I've gone with the current approach; say we at some point decide to switch back to an implementation where Read() is the one doing the actual replacement work - we could then have NewReader() just store the consumedArgs pointer in a private struct member, and have (our new, custom) Read() populate it, with no change to the API.

So moving away from that implementation-dependence was my motivation here.

I will of course add the function comment.